src/shared/dissect-image.c

```diff
@@ -1289,8 +1295,37 @@ static int verity_partition(
 #endif
         } else
                 r = crypt_activate_by_volume_key(cd, name, root_hash, root_hash_size, CRYPT_ACTIVATE_READONLY);
-        if (r < 0)
+        if (r < 0 && r != -EEXIST)
                 return r;
+        if (r == -EEXIST) {
+                /* If the same volume was already open, check that the root hashes match, and reuse it if they do */
+                _cleanup_free_ char *root_hash_existing = NULL;
+                struct crypt_params_verity crypt_params = {};
+                size_t root_hash_existing_size = root_hash_size;
+
+                crypt_free(cd);
+                r = crypt_init_by_name(&cd, name);
+                if (r < 0)
+                        return log_error_errno(r, "Error opening verity device, crypt_init_by_name failed: %m");
+                r = crypt_get_verity_info(cd, &crypt_params);
+                if (r < 0)
+                        return log_error_errno(r, "Error opening verity device, crypt_get_verity_info failed: %m");
+                root_hash_existing = malloc0(root_hash_size);
+                if (!root_hash_existing)
+                        return log_oom();
+                r = crypt_volume_key_get(cd, CRYPT_ANY_SLOT, root_hash_existing, &root_hash_existing_size, NULL, 0);
+                if (r < 0)
+                        return log_error_errno(r, "Error opening verity device, crypt_volume_key_get failed: %m");
+                if (memcmp(root_hash_existing, root_hash, root_hash_size))
+                        return log_error_errno(SYNTHETIC_ERRNO(-EINVAL), "Error opening verity device, it already exists but root hashes are different.");
```

Comment:
SYNTHETIC_ERRNO() uses abs() on the parameter internally, so that it is fine with both negative and positive error numbers. Thus, please drop the `-` here

Reply:
fixed